gateway/api/utils.py

```diff
@@ -165,3 +167,36 @@ def decrypt_env_vars(env_vars: Dict[str, str]) -> Dict[str, str]:
             ) as decryption_error:
                 logger.error("Cannot decrypt %s. %s", key, decryption_error)
     return env_vars
+
+
+def optimistic_lock_model_save(
+    model, max_retries: int = 10, update_fields: Optional[List[str]] = None
+):
+    """Save model while using optimistic lock.
+    Retries if model was not saved successfully.
+
+    Args:
+        model: model to save
+        max_retries: maximum attempt to save model
+        update_fields: fields to update during save
+    """
+    saved = False
+    attempts_left = max_retries
+    while not saved and attempts_left > 0:
+        attempts_left -= 1
+
+        try:
+            if update_fields is not None:
+                model.save(update_fields=update_fields)
+            else:
+                model.save()
+            saved = True
+        except RecordModifiedError:
+            logger.warning(
+                "Model[%s] record has not been updated due to lock. Retrying. Attempts left %s",
+                model,
+                attempts_left,
+            )
+            continue
```

Comment:
I believe that the model record must be refetched and updates re-applied to the fields to resolve `RecordModifiedError` exception.

Reply:
aaaah, good to know!

Reply:
I think that the record need to be refetched in this loop, too.   

My understanding of the this concurrency is that it check if the version of the record in the db is same as the one to be saved.  The version is updated every time the record is saved.  If the version don't match, somebody has updated the the record while the recode is fetched and being saved.

Reply:
I haven't verified but I was wondering if the `version` field need to be added for the `save()` call to make the `concurrency` work.

Reply:
I think I will remove this loop

Reply:
> I haven't verified but I was wondering if the version field need to be added for the save() call to make the concurrency work.

and remove update fields too, I think

Reply:
updated